src/include/duckdb/common/multi_file_reader_function.hpp

```diff
@@ -547,33 +551,70 @@ class MultiFileReaderFunction : public TableFunction {
 		return partition_data;
 	}
 
+	static void InitializeFileScanState(ClientContext &context, MultiFileLocalState &lstate,
+	                                    vector<idx_t> &projection_ids) {
+		auto &reader_data = lstate.reader->reader_data;
+		//! Initialize the intermediate chunk to be used by the underlying reader before being finalized
+		vector<LogicalType> intermediate_chunk_types;
+		auto &local_column_ids = reader_data.column_ids;
+		auto &local_columns = lstate.reader->GetColumns();
+		for (idx_t i = 0; i < local_column_ids.size(); i++) {
+			auto local_idx = MultiFileLocalIndex(i);
+			auto local_id = local_column_ids[local_idx];
+			auto cast_entry = reader_data.cast_map.find(local_id);
+			if (cast_entry == reader_data.cast_map.end()) {
+				auto &col = local_columns[local_id];
+				intermediate_chunk_types.push_back(col.type);
+			} else {
+				intermediate_chunk_types.push_back(cast_entry->second);
+			}
+		}
+		lstate.scan_chunk.Destroy();
+		lstate.scan_chunk.Initialize(context, intermediate_chunk_types);
+
+		auto &executor = lstate.executor;
+		executor.ClearExpressions();
+		if (!projection_ids.empty()) {
+			for (auto &id : projection_ids) {
+				executor.AddExpression(*reader_data.expressions[id]);
+			}
+		} else {
+			for (auto &expr : reader_data.expressions) {
+				executor.AddExpression(*expr);
+			}
+		}
+	}
+
 	static void MultiFileScan(ClientContext &context, TableFunctionInput &data_p, DataChunk &output) {
 		if (!data_p.local_state) {
 			return;
 		}
 		auto &data = data_p.local_state->Cast<MultiFileLocalState>();
 		auto &gstate = data_p.global_state->Cast<MultiFileGlobalState>();
 		auto &bind_data = data_p.bind_data->CastNoConst<MultiFileBindData>();
-		auto remove_columns = gstate.CanRemoveColumns();
-		if (remove_columns) {
-			data.all_columns.Reset();
-		}
+		InitializeFileScanState(context, data, gstate.projection_ids);
```

Comment:
This shouldn't happen for every chunk but should happen for after `TryInitializeBatch` (and maybe then only if the file idx changed)